staging/src/k8s.io/component-base/featuregate/feature_gate_test.go

```diff
@@ -1742,3 +1743,136 @@ func TestResetFeatureValueToDefault(t *testing.T) {
+		expectError bool
+		features    map[Feature]VersionedSpecs
+	}{
+		{
+			name: "normal progression",
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: false, PreRelease: Alpha},
+					{Version: version.MustParse("1.30"), Default: false, PreRelease: Beta},
+					{Version: version.MustParse("1.31"), Default: true, PreRelease: Beta},
+					{Version: version.MustParse("1.32"), Default: true, PreRelease: GA},
+					{Version: version.MustParse("1.33"), Default: false, PreRelease: Deprecated},
+				},
+				testBGate: {
+					{Version: version.MustParse("1.27"), Default: false, PreRelease: Alpha},
+				},
+			},
+		},
+		{
+			name:        "conflicting specs",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testBGate: {
+					{Version: version.MustParse("1.28"), Default: false, PreRelease: Alpha},
+				},
+			},
+		},
+		{
+			name:        "deprecated feature with no prior state",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: false, PreRelease: Deprecated},
+				},
+			},
+		},
+		{
+			name: "deprecated feature with prior state",
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.0"), Default: true, PreRelease: GA},
+					{Version: version.MustParse("1.29"), Default: false, PreRelease: Deprecated},
+				},
+			},
+		},
+		{
+			name:        "duplicate version",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: false, PreRelease: Alpha},
+					{Version: version.MustParse("1.29"), Default: true, PreRelease: Beta},
+				},
+			},
+		},
+		{
+			name:        "decreasing version",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.30"), Default: false, PreRelease: Beta},
+					{Version: version.MustParse("1.29"), Default: true, PreRelease: Beta},
+				},
+			},
+		},
+		{
+			name:        "Beta to Alpha",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: true, PreRelease: Beta},
+					{Version: version.MustParse("1.30"), Default: false, PreRelease: Alpha},
+				},
+			},
+		},
+		{
+			name:        "GA to Alpha",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: true, PreRelease: GA},
+					{Version: version.MustParse("1.30"), Default: false, PreRelease: Alpha},
+				},
+			},
+		},
+		{
+			name:        "GA to Beta",
+			expectError: true,
+			features: map[Feature]VersionedSpecs{
+				testAGate: {
+					{Version: version.MustParse("1.29"), Default: true, PreRelease: GA},
+					{Version: version.MustParse("1.30"), Default: true, PreRelease: Beta},
+				},
+			},
+		},
+		{
+			name: "Alpha to Deprecated to Beta",
```

Comment:
Interesting, I guess this case is allowed but wondering if we have any prior examples?

Reply:
I don't really know. @liggitt @jpbetz should this case be allowed?

Reply:
I'm not aware of any examples.  It would certainly be an exceptional transition.  I think I'd prefer we bias toward not allowing it. This at least forces a discussion if, in the future, anyone thinks they need to do this.

Reply:
it's a weird transition, certainly... I guess I would ask what the invariant we would enforce is... once deprecated, a gate never transitions away from deprecated?

Reply:
if we plan to enforce it, I think the idea is no feature should resurrect from the dead

Reply:
That seems best.  I could imagine asking for a new feature gate if someone wants to resurrect a deprecated thing.  But for now, treating Deprecated as a terminal state seems OK?

Reply:
+1 on treating deprecation as terminal.

Reply:
done

Reply:
I agree with considering deprecation as terminal, but let's be willing to reconsider if we need to.

Deprecation is jargon for "you should not use this"; as a project, we may one day want to change the recommendation. I'm putting this comment in to clarify the story in case it ever comes up.

Imagine that we had a feature, `ZeroContainerPods`. And it went alpha to deprecated because they don't seem useful. One day we come up with a genuine use for zero-container Pods, and we agree (via a KEP?) to un-deprecate the feature. That fictional un-deprecation means we don't have to name the new feature gate `ZeroContainerPodsNoReallyThisTime`.